catalyst/metrics/_cmc_score.py

```diff
@@ -0,0 +1,114 @@
+        for key in shape_type_dict:
+            self.storage[key] = torch.empty(
+                size=shape_type_dict[key]["shape"], dtype=shape_type_dict[key]["dtype"]
+            )
+
+    def update(self, **kwargs) -> None:
+        if self.collected_batches == 0:
+            shape_type_dict = {}
+            for field_name in self.accumulative_fields:
+                shape_type_dict[field_name] = {}
+                shape_type_dict[field_name]["shape"] = (
+                    self.num_samples, *(kwargs[field_name].shape[1:])
+                )
+                shape_type_dict[field_name]["dtype"] = kwargs[field_name].dtype
+            self._allocate_memory(shape_type_dict=shape_type_dict)
+        bs = 0
+        for field_name in self.accumulative_fields:
+            bs = kwargs[field_name].shape[0]
+            self.storage[
+                field_name
+            ][self.tmp_idx: self.tmp_idx + bs, ...] = kwargs[field_name].detach().cpu()
+        self.tmp_idx += bs
+        self.collected_batches += 1
+
+    def compute(self) -> Any:
+        return self.storage
+
+    def compute_key_value(self) -> Dict[str, float]:
+        return self.compute()
+
+
+class CMCMetric(AccumulationMetric):
+    def __init__(
+            self,
+            compute_on_call: bool = True,
+            embeddings_key: str = "features",
+            labels_key: str = "targets",
+            is_query_key: str = "is_query",
+            topk_args: Iterable[int] = (1, ),
+    ):
+        super().__init__(
+            compute_on_call=compute_on_call,
+            accumulative_fields=[embeddings_key, labels_key, is_query_key]
+        )
+        self.embeddings_key = embeddings_key
+        self.labels_key = labels_key
+        self.is_query_key = is_query_key
+        self.topk_args = topk_args
+        self.metric_name = "cmc"
+
+    def compute(self) -> List[float]:
+        query_embeddings = self.storage[
+            self.embeddings_key
+        ][self.storage[self.is_query_key] == 1].float()
+        query_labels = self.storage[self.labels_key][self.storage[self.is_query_key] == 1]
+
+        gallery_embeddings = self.storage[
+            self.embeddings_key
+        ][self.storage[self.is_query_key] == 0].float()
+        gallery_labels = self.storage[self.labels_key][self.storage[self.is_query_key] == 0]
+
+        conformity_matrix = (gallery_labels == query_labels.reshape(-1, 1)).bool()
+
+        metrics = []
+        for k in self.topk_args:
+            value = cmc_score(
+                query_embeddings=query_embeddings,
+                gallery_embeddings=gallery_embeddings,
+                conformity_matrix=conformity_matrix,
+                topk=k,
+            )
+            metrics.append(value)
+
+        return metrics
+
+    def compute_key_value(self) -> Dict[str, float]:
+        values = self.compute()
+        kv_metrics = {
+            f"{self.metric_name}{k:02d}": value for k, value in zip(self.topk_args, values)
```

Comment:
could u please add `prefix` and `suffix` here?
they are required in case when we have a multi-head network and want to compute the metrics for each of them.